src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -167,33 +204,85 @@ alignPaymentSchedule calculationSchedule paymentSchedule = do
 
 -- | Calculate a floating rate amount for each payment date and create claims.
 -- The floating rate is observed according to the FpML ResetDates component and used for the corresponding payment on the last day of that payment period.
-createFloatingRatePaymentClaims : [SchedulePeriod] -> [SchedulePeriod] -> PeriodicSchedule -> Bool -> Decimal -> Bool -> DayCountConventionEnum -> Decimal -> Deliverable -> Observable -> HolidayCalendarData -> SwapStream -> Update TaggedClaim
-createFloatingRatePaymentClaims calculationSchedule paymentSchedule periodicSchedule useAdjustedDatesForDcf floatingRateSpread issuerPays dayCountConvention notional cashInstrumentCid referenceRateId fixingCalendars s = do
+createFloatingRatePaymentClaims : [SchedulePeriod] -> [SchedulePeriod] -> Optional NotionalSchedule -> PeriodicSchedule -> Bool -> Decimal -> Bool -> DayCountConventionEnum -> Deliverable -> Observable -> HolidayCalendarData -> [Date] -> SwapStream -> Party -> Party -> Update [TaggedClaim]
+createFloatingRatePaymentClaims calculationSchedule paymentSchedule constantNotionalScheduleReference periodicSchedule useAdjustedDatesForDcf floatingRateSpread issuerPays dayCountConvention cashInstrumentCid referenceRateId rateFixingCalendars rateFixingDates s issuer calendarDataAgency = do
+  (fxRateId, fxLinkedNotionalInitialValue) <- getFxRateId calculationSchedule paymentSchedule constantNotionalScheduleReference periodicSchedule useAdjustedDatesForDcf floatingRateSpread issuerPays dayCountConvention cashInstrumentCid referenceRateId rateFixingCalendars rateFixingDates s issuer calendarDataAgency
   let
+    fxAdjustmentRequired = isSome fxRateId
+    c = s.calculationPeriodAmount.calculation
+    n = if isSome constantNotionalScheduleReference then fromSome constantNotionalScheduleReference else fromSome c.notionalSchedule
+    notionalSteps = n.notionalStepSchedule.step
+    notionalBase
+      | length notionalSteps == length calculationSchedule
```

Comment:
The same check is repeated in the function above, maybe factor out in common function?

Reply:
This will be done as part of this PR: https://github.com/digital-asset/daml-finance/issues/286
(other repeated code will be removed at the same time).